### PR TITLE
[New Command] add azdev generate-docs

### PR DESCRIPTION
**Pull request by adewaleo**
Work in progress. Add commands to automatically generate CLI reference documentation.

### CODE REVIEW

File: azdev/operations/help/__init__.py [hunk 1]
```diff
@@ -70,3 +76,63 @@ def _help_files_not_in_map(cli_repo, help_files_in_map):
             continue
         not_in_map.append(help_path)
     return not_in_map
+
+def generate_ref_docs(generate_for_extensions=None, output_dir=None, output_type=None):
+    import tempfile
+
+    heading('Generate Reference Docs')
```
Comment (adewaleo) [inline azdev/operations/help/__init__.py]: I will want to probably add more status / progress text

File: azdev/operations/help/__init__.py [hunk 2]
```diff
@@ -70,3 +76,63 @@ def _help_files_not_in_map(cli_repo, help_files_in_map):
...
+        if not os.path.exists(existing_path):
+            raise CLIError("Cannot create output directory {} in non-existent path {}."
+                           .format(base_dir, existing_path))
+
+        os.mkdir(output_dir)
+
+    display("Docs will be placed in {}.".format(output_dir))
+
+    if generate_for_extensions:
+        public_extensions = list_available_extensions()
```
Comment (adewaleo) [inline azdev/operations/help/__init__.py]: Generate for extensions not yet completed

File: azdev/operations/help/refdoc/common/directives.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import copy
+import argparse
+from mock import patch
+from os.path import expanduser
+
+# Directive not in latest release of sphinx, need to pip install sphinx==1.6.7
```
Comment (adewaleo) [inline azdev/operations/help/refdoc/common/directives.py]: As this has been added in setup.py. I should remove this

File: azdev/operations/help/refdoc/common/directives.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    """ Generic Sphinx Directive for generating azure cli documentation.
+        Should be overridden for core CLI or CLI extensions documentation
+    """
+
+    _INDENT = ' ' * 3
+    _DOUBLE_INDENT = _INDENT * 2
+
+    def make_rst(self):
+
+        az_cli = AzCli(cli_name='az',
```
Comment (adewaleo) [inline azdev/operations/help/refdoc/common/directives.py]: todo, refactor these into clearer methods

File: azdev/operations/help/refdoc/common/directives.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+class AbstractHelpGenDirective(Directive):
+    """ Generic Sphinx Directive for generating azure cli documentation.
+        Should be overridden for core CLI or CLI extensions documentation
+    """
+
+    _INDENT = ' ' * 3
+    _DOUBLE_INDENT = _INDENT * 2
+
+    def make_rst(self):
+
```
Comment (adewaleo) [inline azdev/operations/help/refdoc/common/directives.py]: todo: refactor these into smaller methods

File: azdev/operations/help/refdoc/extension_docs/helpgen.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    def _load_doc_source_map(self):
+        # no doc source map for extensions
+        pass
+
+    def _yield_doc_source(self, doc_source_map, help_file):
+        # no doc source map for extensions
+        pass
+
+# TODO: move this into core
```
Comment (adewaleo) [inline azdev/operations/help/refdoc/extension_docs/helpgen.py]: We should really add logic like this into the core, as it is similar to get_all_help

File: azdev/operations/help/refdoc/cli_docs/helpgen.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        here will have no doc source
+    """
+
+    def _get_help_files(self, az_cli):
+        create_invoker_and_load_cmds_and_args(az_cli)
+        return get_all_help(az_cli)
+
+    def _load_doc_source_map(self):
+        return {}
+        # todo: get_cli_repo_path() is failing for some reason.
```
Comment (adewaleo) [inline azdev/operations/help/refdoc/cli_docs/helpgen.py]: @tjprescott get_cli_repo_path() doesn't work for me.

I guess a work around would be for the users to specify doc_source_map's path as an argument to `azdev generate_docs`

Comment (adewaleo) [reply azdev/operations/help/refdoc/cli_docs/helpgen.py]: My b, I think i just needed to re-run azdev setup -c

**Comment by adewaleo** [comment]
Can this be merged? @tjprescott @zikalino 

I think this is good to go

**Comment by tjprescott** [comment]
Thanks for the contribution @adewaleo!

File: azdev/operations/help/__init__.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+from azure.cli.core.extension.operations import list_available_extensions
+
+DOC_MAP_NAME = 'doc_source_map.json'
+HELP_FILE_NAME = '_help.py'
+DOC_SOURCE_MAP_PATH = os.path.join('doc', 'sphinx', 'azhelpgen', DOC_MAP_NAME)
+
+_logger = get_logger(__name__)
+
+
+def check_document_map():
```
[the remark this answers is not in the document]
Comment (adewaleo) [reply azdev/operations/help/__init__.py]: yes it is. I added your recent ado related change too, but please feel free to check if anything is wrong.

File: azdev/operations/help/__init__.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            raise CLIError("Cannot create output directory {} in non-existent path {}."
+                           .format(base_dir, existing_path))
+
+        os.mkdir(output_dir)
+    return output_dir
+
+
+def _generate_ref_docs_for_public_exts(output_type, base_output_dir):
+    # TODO: this shouldn't define the env key, but should reference it from the cli repo.
+    ENV_KEY_AZURE_EXTENSION_DIR = 'AZURE_EXTENSION_DIR'
```
[the remark this answers is not in the document]
Comment (adewaleo) [reply azdev/operations/help/__init__.py]: What I meant is that there isn't a central place this constant is defined in core. I believe there should be.

**Pull request merged**